src/sys/sgx/tcp/stream.rs

```diff
@@ -0,0 +1,488 @@
+    connect_state: State<String, Option<CancelHandle>, net::TcpStream>,
+    #[cfg(compiler_has_send_sgx_types)]
+    write_buffer: WriteBuffer,
+    #[cfg(not(compiler_has_send_sgx_types))]
+    write_buffer: MakeSend<WriteBuffer>,
+    write_state: State<(), Option<CancelHandle>, ()>,
+    #[cfg(compiler_has_send_sgx_types)]
+    read_buf: Option<User<[u8]>>,
+    #[cfg(not(compiler_has_send_sgx_types))]
+    read_buf: Option<MakeSend<User<[u8]>>>,
+    #[cfg(compiler_has_send_sgx_types)]
+    read_state: State<(), Option<CancelHandle>, ReadBuffer>,
+    #[cfg(not(compiler_has_send_sgx_types))]
+    read_state: State<(), Option<CancelHandle>, MakeSend<ReadBuffer>>,
+    registration: Option<Registration>,
+    provider: Option<Provider>,
+}
+
+impl TcpStream {
+    fn new(connect_state: State<String, Option<CancelHandle>, net::TcpStream>) -> Self {
+        TcpStream {
+            imp: StreamImp(Arc::new(Mutex::new(StreamInner {
+                connect_state,
+                #[cfg(compiler_has_send_sgx_types)]
+                write_buffer: WriteBuffer::new(User::<[u8]>::uninitialized(WRITE_BUFFER_SIZE)),
+                #[cfg(not(compiler_has_send_sgx_types))]
+                write_buffer: MakeSend::new(WriteBuffer::new(User::<[u8]>::uninitialized(WRITE_BUFFER_SIZE))),
+                write_state: State::New(()),
+                #[cfg(compiler_has_send_sgx_types)]
+                read_buf: Some(User::<[u8]>::uninitialized(READ_BUFFER_SIZE)),
+                #[cfg(not(compiler_has_send_sgx_types))]
+                read_buf: Some(MakeSend::new(User::<[u8]>::uninitialized(READ_BUFFER_SIZE))),
+                read_state: State::New(()),
+                registration: None,
+                provider: None,
+            }))),
+        }
+    }
+
+    pub(super) fn from_std(stream: net::TcpStream) -> (Self, SocketAddr) {
+        let peer_addr = stream.peer_addr().unwrap_or_else(|_| ([0; 4], 0).into());
+        let stream = TcpStream::new(State::Ready(stream));
+        (stream, peer_addr)
+    }
+
+    pub fn connect(addr: SocketAddr) -> io::Result<TcpStream> {
+        Ok(TcpStream::new(State::New(addr.to_string())))
+    }
+
+    pub fn connect_str(addr: &str) -> io::Result<TcpStream> {
+        Ok(TcpStream::new(State::New(addr.to_owned())))
+    }
+
+    pub fn peer_addr(&self) -> io::Result<SocketAddr> {
+        self.inner()
+            .connect_state
+            .as_ready()
+            .ok_or_else(|| would_block())
+            .and_then(|stream| stream.peer_addr())
+    }
+
+    pub fn local_addr(&self) -> io::Result<SocketAddr> {
+        self.inner()
+            .connect_state
+            .as_ready()
+            .ok_or_else(|| would_block())
+            .and_then(|stream| stream.local_addr())
+    }
+
+    pub fn shutdown(&self, _how: Shutdown) -> io::Result<()> {
+        Ok(()) // ineffective in SGX
+    }
+
+    pub fn set_nodelay(&self, _nodelay: bool) -> io::Result<()> {
+        Ok(()) // ineffective in SGX
```

Comment:
Isn't it important to implement `nodelay` for performance?

Reply:
It can be useful, but we don't have a way to configure this in SGX.